pkg/antctl/raw/packetcapture/command_test.go

```diff
@@ -0,0 +1,274 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packetcapture
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/kubernetes"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+	"k8s.io/client-go/rest"
+	k8stesting "k8s.io/client-go/testing"
+	"k8s.io/utils/ptr"
+
+	"antrea.io/antrea/pkg/antctl/raw"
+	"antrea.io/antrea/pkg/apis/crd/v1alpha1"
+	antrea "antrea.io/antrea/pkg/client/clientset/versioned"
+	antreafakeclient "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+)
+
+const (
+	srcPod            = "default/pod-1"
+	dstPod            = "pod-2"
+	ipv4              = "192.168.10.10"
+	testNum     int32 = 10
+	testTimeout       = 3 * time.Second
+)
+
+var (
+	antreaAgentPod = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "antrea-agent-1",
+			Namespace: "kube-system",
+		},
+	}
+	pod1 = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod-1",
+			Namespace: "default",
+		},
+	}
+	pod2 = v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "pod-2",
+			Namespace: "default",
+		},
+	}
+	k8sClient = k8sfake.NewSimpleClientset(&pod1, &pod2, &antreaAgentPod)
+)
+
+type testPodFile struct {
+}
+
+func (p *testPodFile) CopyFromPod(ctx context.Context, fs afero.Fs, namespace, name, containerName, srcPath, dstDir string) error {
+	return nil
+}
+
+func setCommandOptions(src, dst, notwait, flow string) {
+	Command.Flags().Set("source", src)
+	Command.Flags().Set("destination", dst)
+	Command.Flags().Set("nowait", notwait)
+	Command.Flags().Set("number", "10")
+	Command.Flags().Set("flow", flow)
+}
+
+func TestRun(t *testing.T) {
+	tcs := []struct {
+		name      string
+		src       string
+		dst       string
+		nowait    string
+		flow      string
+		expectErr string
+	}{
+		{
+			name: "pod-2-pod",
+			src:  srcPod,
+			dst:  dstPod,
+			flow: "tcp,tcp_src=50060,tcp_dst=80",
+		},
+		{
+			name:   "pod-2-ip",
+			src:    srcPod,
+			dst:    ipv4,
+			nowait: "true",
+			flow:   "udp,udp_src=1234,udp_dst=1234",
+		},
+		{
+			name:      "timeout",
+			src:       srcPod,
+			dst:       dstPod,
+			flow:      "icmp",
+			expectErr: "timeout while waiting for PacketCapture to complete",
+		},
+		{
+			name:      "invalid-packetcapture",
+			src:       ipv4,
+			dst:       ipv4,
+			flow:      "icmp",
+			expectErr: "error when constructing a PacketCapture CR: one of source and destination must be a Pod",
+		},
+	}
+
+	for _, tt := range tcs {
+		t.Run(tt.name, func(t *testing.T) {
+			setCommandOptions(tt.src, tt.dst, tt.nowait, tt.flow)
```

Comment:
rather than testing `packetCaptureRunE`, I feel like we should put all the logic in a different function which would be easier to test. That would make the tests cleaner. What I mean is we could have this:

```go
func packetCaptureRunE(cmd *cobra.Command, args []string) error {
	option.timeout, _ = cmd.Flags().GetDuration("timeout")
        return packetCaptureRun(cmd.Context(), options)
}

func packetCaptureRun(ctx context.Context, options type_needed) error {
        // ...
}
```

And then have unit tests for  `packetCaptureRun`.

Reply:
updated. Thanks